core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -4537,28 +4537,22 @@ object Stream extends StreamLowPriority {
     private[fs2] def apply[F[_]: Temporal, O, O2, R](source: Stream[F, O])(pull: TimedPull[F, O] => Pull[F, O2, R]): Pull[F, O2, R] = {
       def now = Temporal[F].monotonic
 
-      class Timeout(val id: Token, issuedAt: FiniteDuration, d: FiniteDuration) {
-        def asOfNow:  F[FiniteDuration] = now.map(now => d - (now - issuedAt))
-      }
+      case class Timeout(id: Token, d: FiniteDuration)
 
       def newTimeout(d: FiniteDuration): F[Timeout] =
-        (Token[F], now).mapN(new Timeout(_, _, d))
+        Token[F].map(new Timeout(_, d))
 
       Pull
         .eval { newTimeout(0.millis) mproduct SignallingRef.of[F, Timeout] }
         .flatMap { case (initial, time) =>
-          def nextAfter(t: Timeout): Stream[F, Nothing] =
-            time.discrete.dropWhile(_.id == t.id).head.drain
 
           def timeouts: Stream[F, Token] =
-            nextAfter(initial) ++
-           Stream
-            .eval { time.get.mproduct(_.asOfNow) }
-            .flatMap { case (currentTimeout, timeToWait) =>
-              if (timeToWait <= 0.nanos)
-                Stream.emit(currentTimeout.id) ++ nextAfter(currentTimeout)
-              else Stream.sleep_[F](timeToWait)
-            }.repeat
+            time
+              .discrete
+              .dropWhile(_.id == initial.id)
+              .switchMap { timeout =>
+                Stream.sleep(timeout.d).as(timeout.id)
+              }
```

Comment:
I have a question on the commit here @mpilquist @djspiewak . I think only data can give a definite answer but I just want to gauge what you think first. Look at the diff for implementing `timeouts` in that commit. 
The problem is implementing resettable timeouts.
The first one only uses one extra fiber, but it only works if  calls to `resetTimer` never try to reset a timeout which hasn't triggered yet with a shorter one.
The second is fully general, but it switches fiber at every timeout reset.

My question is, do you think the first (based on your sensibility) would offer a significant advantage in perf and/or memory utilisation? If so, do you think it's worth trying to introduce it as a "fast-path" for when `resetTimer` is called in the right conditions (and only fallback to `switchMap` if not), or is that too much trouble for little gain?

Reply:
Interesting question. Pondering…

One thought here is that fibers are extremely cheap. The context shift around `start` doesn't even involve memory barriers by default, since it just plops the new fiber onto the same worker thread, so you can pretty safely spam that function. The allocation itself creates some memory pressure, but the resident size of `IOFiber` is not enormous. It's basically analogous to the magnitude of allocations you need to perform if you interpret `.pull.uncons1` on a small (~16) chunk, but with fewer objects. Specifically, five objects get allocated in addition to the `IOFiber` when you call `start`, two of which are small arrays. The object header of `IOFiber` itself is the largest memory footprint in the whole menagerie.

Tldr, I suspect that vomiting out fibers is going to be fine. It's not something I would want to be doing in a hot path, but I also wouldn't want to be hitting `Stream` chunk boundaries within a hot path either (for the same reason), and it's around the same order of magnitude cost.

Reply:
Cool. Just note that `switchMap` is a lot more than _just_ creating a Fiber (queues, semaphores etc are involved), but I think the memory pressure concern would be the same, and sounds like it shouldn't be too much of an issue